src/Draco.Chr/Solve/PrioritizingSolver.cs

```diff
@@ -0,0 +1,178 @@
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using Draco.Chr.Constraints;
+using Draco.Chr.Rules;
+using Draco.Chr.Tracing;
+
+namespace Draco.Chr.Solve;
+
+/// <summary>
+/// A base class for solvers that only need to implement rule prioritization.
+/// </summary>
+public abstract class PrioritizingSolver(
+    IEqualityComparer? comparer = null,
+    ITracer? tracer = null) : ISolver
+{
+    private readonly record struct RuleAndMatch(Rule Rule, ImmutableArray<IConstraint> Match);
+
+    /// <summary>
+    /// The rules in priority order.
+    /// </summary>
+    public abstract IEnumerable<Rule> Rules { get; }
```

Comment:
Why would we need different implementations ?

Reply:
It's some really simple future proofing in case we want to be able to define different rule application orders. This would be the case, when we'd want to improve the performance of applicable rule search (tho that'd also involve some interface changes for sure).